### PR TITLE
[#3117]Just Sink the Notify implementation into common module and optimize some parts

### PR DESCRIPTION
**Pull request by zongtanghu**
Please do not create a Pull Request without creating an issue first.

## What is the purpose of the change

 Sink the Notify implementation into common module and optimize some parts

## Brief changelog

 Sink the Notify implementation into common module and optimize some parts.
And this pr is one part of another bigger pr.The link is #2859.

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

* [x] Make sure there is a Github issue filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
* [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
* [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
* [x] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/alibaba/nacos/tree/master/test).
* [x] Run `mvn -B clean package apache-rat:check findbugs:findbugs -Dmaven.test.skip=true` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.

### CODE REVIEW

**Comment by zongtanghu** [comment]
这个PR的几个改造要点：

（1）：将原来core module里面的NofityCenter和DefaultPublisher等下沉至common的module中，同时由于common module本身只是支持 java 6的，所以我把原来Nofity里面涉及Lamba表达式和Java 8的东西，都替换掉了（其中为了去除原本java 8的default关键字，我用了abstract抽象类来代替；将BiFunction从JDK 8中剥离移到common module中），没有做实际的重构和改造，基本是保留原来春少的Nofity设计思路的；

（2）：对于“System.currentTimeMillis()”设置sequence的方法，采用“AtomicInteger或者AtomicLong，减少对系统时间的查询”

**Comment by zongtanghu** [comment]
Please help to review this pr, @yanlinly @KomachiSion @chuntaojun

File: common/src/main/java/com/alibaba/nacos/common/utils/MapUtils.java [hunk 1]
```diff
@@ -124,4 +124,31 @@ public static void putIfValNoEmpty(Map target, Object key, Object value) {
...
+     * ComputeIfAbsent lazy load.
+     *
+     * @param target target Map data.
+     * @param key map key.
+     * @param mappingFunction funtion which is need to be executed.
+     * @param param1 function's parameter value1.
+     * @param param2 function's parameter value1.
+     * @return
+     */
+    public static Object computeIfAbsent(Map target, Object key, BiFunction mappingFunction, Object param1,
```
Comment (chuntaojun) [inline common/src/main/java/com/alibaba/nacos/common/utils/MapUtils.java:137]: 标注下，非线程安全的操作

File: common/src/main/java/com/alibaba/nacos/common/utils/ClassUtils.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+/**
+ * Utils for Class.
+ *
+ * @author <a href="mailto:liaochuntao@live.com">liaochuntao</a>
+ */
+@SuppressWarnings("all")
+public final class ClassUtils {
+    
+    public static Class findClassByName(String className) {
```
[the remark this answers is not in the document]
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/utils/ClassUtils.java:27]: Okay, I will add java doc.

File: common/src/main/java/com/alibaba/nacos/common/utils/BiFunction.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    /**
+     * Applies this function to the given arguments.
+     *
+     * @param t the first function argument
+     * @param u the second function argument
+     * @return the function result
+     */
+    R apply(T t, U u);
+
+    //    end
```
[the remark this answers is not in the document]
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/utils/BiFunction.java:42]: This comments represent this part is copy from jdk source codes.

File: common/src/main/java/com/alibaba/nacos/common/utils/BiFunction.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ * <p>This is a <a href="package-summary.html">functional interface</a>
+ * whose functional method is {@link #apply(Object, Object)}.
+ *
+ * @author zongtanghu
+ *
+ */
+public interface BiFunction<T, U, R> {
+
+    //    The following utility functions are extracted from <link>org.apache.commons.lang3</link>
+    //    start
```
[the remark this answers is not in the document]
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/utils/BiFunction.java:31]: This comments represent this part is copy from jdk source codes.

File: common/src/main/java/com/alibaba/nacos/common/notify/Event.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ * @author <a href="mailto:liaochuntao@live.com">liaochuntao</a>
+ * @author zongtanghu
+ */
+@SuppressWarnings("all")
+public abstract class Event implements Serializable {
+    
+    private static final AtomicLong SEQUENCE = new AtomicLong(0);
+    
+    private long no = SEQUENCE.getAndIncrement();
```
[the remark this answers is not in the document]
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/notify/Event.java:33]: Okay I will fix this in next commit.

File: common/src/main/java/com/alibaba/nacos/common/notify/Event.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import java.io.Serializable;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * An abstract class for event.
+ *
+ * @author <a href="mailto:liaochuntao@live.com">liaochuntao</a>
+ * @author zongtanghu
+ */
+@SuppressWarnings("all")
```
[the remark this answers is not in the document]
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/notify/Event.java:28]: Okay I will fix this in next commit.

File: common/src/main/java/com/alibaba/nacos/common/notify/DefaultPublisher.java [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+        return CollectionUtils.isNotEmpty(subscribers) || CollectionUtils.isNotEmpty(SMART_SUBSCRIBERS);
+    }
+    
+    @Override
+    public void addSubscriber(Subscriber subscriber) {
+        subscribers.add(subscriber);
+    }
+    
+    @Override
+    public void unSubscriber(Subscriber subscriber) {
```
[the remark this answers is not in the document]
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/notify/DefaultPublisher.java:134]: Okay I will fix this in next commit.

File: common/src/main/java/com/alibaba/nacos/common/utils/BiFunction.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/utils/BiFunction.java:31]: Okay I will fix this in next commit.

File: common/src/main/java/com/alibaba/nacos/common/utils/BiFunction.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/utils/BiFunction.java:42]: Okay I will fix this in next commit.

File: common/src/main/java/com/alibaba/nacos/common/utils/MapUtils.java (hunk 1 above)
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/utils/MapUtils.java:137]: Okay I will fix this in next commit.

File: common/src/main/java/com/alibaba/nacos/common/utils/ClassUtils.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        try {
+            return Class.forName(className);
+        } catch (Exception e) {
+            throw new RuntimeException("this class name not found");
```
[the remark this answers is not in the document]
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/utils/ClassUtils.java:31]: Okay I will fix this in next commit.

File: common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java [hunk 9]
```diff
@@ -86,19 +89,23 @@
         BUILD_FACTORY = new BiFunction<Class<? extends Event>, Integer, EventPublisher>() {
             
             @Override
-            public EventPublisher apply(Class<? extends Event> cls, Integer buffer) {
+            public EventPublisher apply(Class<? extends Event> cls, Integer buffer) throws NacosException {
```
Comment (Maijh97) [inline common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java]: 目前我发现在另外的一个pr #3114 提及会从nacos-common模块中移除nacos-api模块，`NacosException`是所属nacos-api模块的。这里是不是可以用`RuntimeException `代替。

File: common/src/main/java/com/alibaba/nacos/common/utils/ClassUtils.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (Maijh97) [reply common/src/main/java/com/alibaba/nacos/common/utils/ClassUtils.java]: 目前我发现在另外的一个pr #3114 提及会从nacos-common模块中移除nacos-api模块，NacosException是所属nacos-api模块的。这里是不是可以用RuntimeException 代替。

File: common/src/main/java/com/alibaba/nacos/common/NotThreadSafe.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+/**
+ * Annotation that marks a method as not thread safe.
+ *
+ * @author zongtanghu
+ */
+@Documented
+@Target({ElementType.TYPE, ElementType.METHOD})
+@Retention(RetentionPolicy.SOURCE)
+public @interface NotThreadSafe {
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/NotThreadSafe.java:33]: I know you want to figure out some method may be not thread safe. But I want to know whether is project do like this? I think write some discription in Javadoc is enough.

File: common/src/main/java/com/alibaba/nacos/common/notify/Event.java [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+/**
+ * An abstract class for event.
+ *
+ * @author <a href="mailto:liaochuntao@live.com">liaochuntao</a>
+ * @author zongtanghu
+ */
+@SuppressWarnings({"PMD.AbstractClassShouldStartWithAbstractNamingRule", "PMD.ConstantFieldShouldBeUpperCaseRule"})
+public abstract class Event implements Serializable {
+    
+    private static final AtomicLong sequence = new AtomicLong(0);
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/notify/Event.java:31]: I mean that  `no-->sequence`  the constant still is SEQUENCE.

File: common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+
+import static com.alibaba.nacos.api.exception.NacosException.SERVER_ERROR;
+
+/**
+ * Unified Event Notify Center.
+ *
+ * @author <a href="mailto:liaochuntao@live.com">liaochuntao</a>
+ * @author zongtanghu
+ */
+@SuppressWarnings("PMD.ConstantFieldShouldBeUpperCaseRule")
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java:46]: Why you add this SuppressWarnings

File: common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+public class NotifyCenter {
+    
+    private static final Logger LOGGER = LoggerFactory.getLogger(NotifyCenter.class);
+    
+    public static int RING_BUFFER_SIZE = 16384;
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java:51]: No constants variable, use lower.

File: common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+    
+    public static int SHARE_BUFFER_SIZE = 1024;
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java:53]: No constants variable, use lower.

File: common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java [hunk 15]
```diff
@@ -0,0 +1,297 @@
...
+    
+    private static final AtomicBoolean CLOSED = new AtomicBoolean(false);
+    
+    private static BiFunction<Class<? extends Event>, Integer, EventPublisher> BUILD_FACTORY = null;
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java:57]: No constants variable, use lower.

File: common/src/main/java/com/alibaba/nacos/common/utils/ClassUtils.java [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+
+import static com.alibaba.nacos.api.exception.NacosException.SERVER_ERROR;
+
+
+/**
+ * Utils for Class.
+ *
+ * @author <a href="mailto:liaochuntao@live.com">liaochuntao</a>
+ */
+@SuppressWarnings("all")
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/utils/ClassUtils.java:29]: Remove SuppressWarnings

File: common/src/main/java/com/alibaba/nacos/common/utils/BiFunction.java [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+ * Represents a function that accepts two arguments and produces a result.
+ * The following utility functions are extracted from org.apache.commons.lang3.
+ *
+ * <p>This is a <a href="package-summary.html">functional interface</a>
+ * whose functional method is {@link #apply(Object, Object)}.
+ *
+ * @author zongtanghu
+ *
+ */
+@SuppressWarnings("PMD.AbstractMethodOrInterfaceMethodMustUseJavadocRule")
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/utils/BiFunction.java:31]: Remove SuppressWarnings.

File: common/src/main/java/com/alibaba/nacos/common/notify/listener/Subscriber.java [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+
+import java.util.concurrent.Executor;
+
+/**
+ * An abstract subscriber class for subscriber interface.
+ *
+ * @author <a href="mailto:liaochuntao@live.com">liaochuntao</a>
+ * @author zongtanghu
+ */
+@SuppressWarnings("all")
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/notify/listener/Subscriber.java:29]: remove SuppressWarnings

File: common/src/main/java/com/alibaba/nacos/common/NotThreadSafe.java (hunk 10 above)
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/NotThreadSafe.java:33]: Some open source project use this annotation to present unsafe method.

File: common/src/main/java/com/alibaba/nacos/common/notify/Event.java (hunk 11 above)
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/notify/Event.java:31]: Okay, sorry , I disunderstand what's your meaning.

Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/notify/Event.java:31]: I will fix this in next commit.

File: common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java (hunk 13 above)
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java:51]: I will fix this in next commit.

File: common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java (hunk 14 above)
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java:53]: I will fix this in next commit.

File: common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java (hunk 15 above)
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java:57]: I will fix this in next commit.

File: common/src/main/java/com/alibaba/nacos/common/utils/ClassUtils.java (hunk 16 above)
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/utils/ClassUtils.java:29]: I will fix this in next commit.

File: common/src/main/java/com/alibaba/nacos/common/utils/BiFunction.java (hunk 17 above)
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/utils/BiFunction.java:31]: I will fix this in next commit.

File: common/src/main/java/com/alibaba/nacos/common/notify/listener/Subscriber.java (hunk 18 above)
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/notify/listener/Subscriber.java:29]: I will fix this in next commit.

File: common/src/main/java/com/alibaba/nacos/common/NotThreadSafe.java (hunk 10 above)
Comment (KomachiSion) [reply common/src/main/java/com/alibaba/nacos/common/NotThreadSafe.java:33]: OK

File: common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java [hunk 19]
```diff
@@ -0,0 +1,331 @@
...
+     *
+     * @param consumer subscriber
+     * @param <T>      event type
+     */
+    public static <T> void registerSubscriber(final Subscriber consumer) throws NacosException {
+        final Class<? extends Event> cls = consumer.subscribeType();
+        // If you want to listen to multiple events, you do it separately,
+        // based on subclass's subscribeTypes method return list, it can register to publisher.
+        if (consumer instanceof SmartSubscriber) {
+            EventPublisher.SMART_SUBSCRIBERS.add((SmartSubscriber) consumer);
```
Comment (chuntaojun) [inline common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java:179]: delete SMART_SUBSCRIBERS

File: common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java [hunk 20]
```diff
@@ -0,0 +1,331 @@
...
+    
+    /**
+     * Deregister subscriber.
+     *
+     * @param consumer subscriber instance.
+     */
+    public static <T> void deregisterSubscriber(final Subscriber consumer) {
+        final Class<? extends Event> cls = consumer.subscribeType();
+        if (consumer instanceof SmartSubscriber) {
+            EventPublisher.SMART_SUBSCRIBERS.remove((SmartSubscriber) consumer);
```
Comment (chuntaojun) [inline common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java:218]: delete SMART_SUBSCRIBERS

File: common/src/main/java/com/alibaba/nacos/common/notify/DefaultPublisher.java [hunk 21]
```diff
@@ -0,0 +1,223 @@
...
+                receiveEvent(event);
+                updater.compareAndSet(this, lastEventSequence, Math.max(lastEventSequence, event.sequence()));
+            }
+        } catch (Throwable ex) {
+            LOGGER.error("Event listener exception : {}", ex);
+        }
+    }
+    
+    private boolean hasSubscriber() {
+        return CollectionUtils.isNotEmpty(subscribers) || CollectionUtils.isNotEmpty(SMART_SUBSCRIBERS);
```
[the remark this answers is not in the document]
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/notify/DefaultPublisher.java]: Okay, this issue has already been resolved in this pr.

File: common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java (hunk 19 above)
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java:179]: I will remove this in next commit.

File: common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java (hunk 20 above)
Comment (zongtanghu) [reply common/src/main/java/com/alibaba/nacos/common/notify/NotifyCenter.java:218]: I will remove this in next commit.

**Comment by zongtanghu** [comment]
Hi @yanlinly， 我再下一个pr里面把ut一起迁移到common库下面吧



huzongtang@cmss.chinamobile.com
 
From: yanlinly
Date: 2020-06-22 10:29
To: alibaba/nacos
CC: Hu Zongtang; Author
Subject: Re: [alibaba/nacos] [#3117]Just Sink the Notify implementation into common module and optimize some parts (#3118)
@yanlinly commented on this pull request.
后面建议把ut 也一起迁移下来。
―

**Pull request merged**